### ISSUE TITLE
`cargo install` ignores --index and --registry when downloading package dependencies

### ISSUE THREAD

**Issue by mikhasd** [rust-lang/cargo#12076]
### Problem

When running `cargo install --index sparse+https://INTERNAL grcov` or `cargo install --registry INTERNAL grcov`, cargo will download the first package from the specified registry/index and uses crates.io for it's dependencies.

### Steps

```bash
export CARGO_REGISTRIES_INTERNAL_INDEX=sparse+https://INTERNAL/index
export CARGO_REGISTRY_DEFAULT=INTERNAL
cargo install grcov
# or
cargo install --index sparse+https://INTERNAL/index grcov
# or
cargo install --registry INTERNAL grcov
```
Outputs:
```
    Updating `sparse+https://INTERNAL/index/` index
  Installing grcov v0.8.18 (registry `sparse+https://INTERNAL/index/`)
    Updating crates.io index
warning: spurious network error (2 tries remaining): [28] Timeout was reached (Failed to connect to github.com port 443 after 21004 ms: Couldn't connect to server); class=Net (12)
warning: spurious network error (1 tries remaining): [28] Timeout was reached (Failed to connect to github.com port 443 after 21001 ms: Couldn't connect to server); class=Net (12)
error: failed to compile `grcov v0.8.18 (registry `sparse+https://INTERNAL/index/`)`, intermediate artifacts can be found at `C:\Users\XXXXXXXXX\AppData\Local\Temp\1\cargo-installHaebDK`

Caused by:
  failed to get `cargo-binutils` as a dependency of package `grcov v0.8.18 (registry `sparse+https://INTERNAL/index/`)`

Caused by:
  failed to load source for dependency `cargo-binutils`

Caused by:
  Unable to update registry `crates-io`

Caused by:
  failed to fetch `https://github.com/rust-lang/crates.io-index`

Caused by:
  network failure seems to have happened
  if a proxy or similar is necessary `net.git-fetch-with-cli` may help here
  https://doc.rust-lang.org/cargo/reference/config.html#netgit-fetch-with-cli

Caused by:
  [28] Timeout was reached (Failed to connect to github.com port 443 after 21001 ms: Couldn't connect to server); class=Net (12)
```

### Notes

I assumed cargo would use the provided registry/index configuration to download the desired package as well as its dependencies.

### Version

```text
cargo 1.69.0 (6e9a83356 2023-04-12)
```

**Comment by mikhasd** [comment]
That was my initial thoughts, but after download the package index file, I found no references to crates.io on it: https://index.crates.io/gr/co/grcov.

Notice that the behavior I expected for `install` is the one I see today with `cargo add` or `build`: these commands downloads the package and its dependencies from the specifying the index or registry.

**Comment by mikhasd** [comment]
> In an index, when the index is not specified cargo interprets it as "from the same index". That is why the package index file on crates.io does not need to mention crates.io.

On my mirror/internal index, the index is not specified as well, so it should download everything from my internal, which is not happening.

**Comment by ehuss** [comment]
Oh, I see what is happening.

`cargo install` just consults the registry to download the package that is passed on the command-line. Cargo then extracts that `.crate` file into a temporary directory and runs a normal `cargo build` on it, forgetting about the index that it came from. For packages on crates.io, their `Cargo.toml` dependencies do not specify a registry, so all the dependencies come from crates.io.

I still think the correct solution is to use source replacement to implement mirroring. Cargo isn't really designed for supporting mirroring with the `--index/--registry` flags.

**Comment by mikhasd** [comment]
Thanks for the explanation.

So, from an enterprise perspective, the best approach to make developers use an internal registry/index by default is to use source replacement, configured on their $CARGO_HOME/config.toml(?).

**Comment by Eh2406** [comment]
> So, from an enterprise perspective, the best approach to make developers use an internal registry/index by default is to use source replacement, configured on their $CARGO_HOME/config.toml(?).

Yes.

> forgetting about the index that it came from. For packages on crates.io, their Cargo.toml dependencies do not specify a registry, so all the dependencies come from crates.io.

That really sounds like a bug. If the registry is not a mirror, then this will ask crates.io for dependencies that are only available from the internal registry, even though the package the publish came from clearly stated that the dependency should come from the internal registry.

**Comment by ehuss** [comment]
> If the registry is not a mirror, then this will ask crates.io for dependencies that are only available from the internal registry, even though the package the publish came from clearly stated that the dependency should come from the internal registry.

Wouldn't those crates have dependencies with the `registry="…"` field in `Cargo.toml`?

**Comment by ehuss** [comment]
Note that there are other issues like #11406 about making it easier to support mirroring in various ways, particularly with a single registry that is a "front" for both internal and external registries.

**Comment by mikhasd** [comment]
I agree with the previous comments, but cargo install not respecting CARGO_REGISTRY_DEFAULT feels like an incorrect behavior.
